### PR TITLE
More tests for docker integration

### LINKED ISSUE #237: More test for docker tagger.

### PR DESCRIPTION
**Pull request by tomwilkie**
Fixes #237

- Move docker probe code into it's own module
- Put PIDTree behind an interface for mocking
- Disaggregate dockerTagger into a registry, tagger and reporter
- Similarly disaggregate tests
- Add mocks for docker container and registry
- Add test for docker events.

### CODE REVIEW

**Comment by tomwilkie** [comment]
Note the inclusion of work from #254; this will go away when that is merged.

File: probe/docker/registry.go [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+		quit:     make(chan struct{}),
+	}
+
+	go r.loop()
+	return r, nil
+}
+
+// Stop stops the Docker tagger's event subscriber.
+func (r *registry) Stop() {
+	close(r.quit)
```
Comment (paulbellamy) [inline probe/docker/registry.go]: Would be nice to wait for this to quit.

File: probe/docker/registry.go [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+	ticker := time.Tick(r.interval)
+	for {
+		select {
+		case <-ticker:
+			if !r.update() {
+				return
+			}
+
+		case <-r.quit:
+			return
```
Comment (paulbellamy) [inline probe/docker/registry.go]: This should shutdown the connections, probably?

File: probe/docker/registry.go [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+		}
+	}
+}
+
+func (r *registry) update() bool {
```
Comment (paulbellamy) [inline probe/docker/registry.go]: You're locking/unlocking the registry a bunch during this process, which could lead to queries returning half-complete data.

File: probe/docker/reporter.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+func NewReporter(registry Registry, scope string) *Reporter {
+	return &Reporter{
+		registry: registry,
+		scope:    scope,
+	}
+}
+
+// Report generates a Report containing Container and ContainerImage topologies
+func (r *Reporter) Report() report.Report {
+	result := report.MakeReport()
```
Comment (paulbellamy) [inline probe/docker/reporter.go]: maybe `report.New()`?

File: probe/main.go [hunk 5]
```diff
@@ -73,18 +74,21 @@ func main() {
...
+
 	taggers := []tag.Tagger{tag.NewTopologyTagger(), tag.NewOriginHostTagger(hostID)}
+	reporters := []tag.Reporter{}
+
 	if *dockerEnabled && runtime.GOOS == linux {
-		var err error
-		dockerTagger, err = tag.NewDockerTagger(*procRoot, *dockerInterval)
+		dockerRegistry, err := docker.NewRegistry(*dockerInterval)
 		if err != nil {
 			log.Fatalf("failed to start docker tagger: %v", err)
```
Comment (paulbellamy) [inline probe/main.go]: Do you mean "registry" here?

File: probe/tag/origin_host_tagger.go [hunk 6]
```diff
@@ -12,12 +12,12 @@ func NewOriginHostTagger(hostID string) Tagger {
 	return &originHostTagger{hostNodeID: report.MakeHostNodeID(hostID)}
 }
 
-func (t originHostTagger) Tag(r report.Report) report.Report {
+func (t originHostTagger) Tag(r report.Report) (report.Report, error) {
```
Comment (paulbellamy) [inline probe/tag/origin_host_tagger.go]: Why return error, if we always return nil?

File: probe/tag/origin_host_tagger_test.go [hunk 7]
```diff
@@ -18,7 +18,8 @@ func TestOriginHostTagger(t *testing.T) {
 	r := report.MakeReport()
 	r.Endpoint.NodeMetadatas[endpointNodeID] = nodeMetadata
 	want := nodeMetadata.Merge(report.NodeMetadata{report.HostNodeID: report.MakeHostNodeID(hostID)})
-	have := tag.NewOriginHostTagger(hostID).Tag(r).Endpoint.NodeMetadatas[endpointNodeID].Copy()
+	rpt, _ := tag.NewOriginHostTagger(hostID).Tag(r)
```
Comment (paulbellamy) [inline probe/tag/origin_host_tagger_test.go]: This can't fail?

File: probe/tag/topology_tagger.go [hunk 8]
```diff
@@ -11,7 +11,7 @@ func NewTopologyTagger() Tagger {
 	return &topologyTagger{}
 }
 
-func (topologyTagger) Tag(r report.Report) report.Report {
+func (topologyTagger) Tag(r report.Report) (report.Report, error) {
```
Comment (paulbellamy) [inline probe/tag/topology_tagger.go]: Again, if this can't fail, why return an error?

File: probe/tag/origin_host_tagger_test.go (hunk 7 above)
Comment (paulbellamy) [inline probe/tag/origin_host_tagger_test.go]: Ah, right. this implementation of the interface can't fail. Fine.

File: probe/docker/registry.go (hunk 1 above)
Comment (tomwilkie) [inline probe/docker/registry.go]: Yes, good point.  Done.

File: probe/docker/registry.go (hunk 2 above)
Comment (tomwilkie) [inline probe/docker/registry.go]: Agreed.  Done.

File: probe/docker/registry.go (hunk 3 above)
Comment (tomwilkie) [inline probe/docker/registry.go]: Yes its unfortunate, but all the monitoring in scope is best-effort, and incomplete data is acceptable in this case.

File: probe/docker/reporter.go (hunk 4 above)
Comment (tomwilkie) [inline probe/docker/reporter.go]: That would be nice, but I'll leave that up to @peterbourgon and a future change.  Peter is a stickler for naming.

File: probe/main.go (hunk 5 above)
Comment (tomwilkie) [inline probe/main.go]: Ah yes!

File: probe/tag/origin_host_tagger_test.go (hunk 7 above)
Comment (tomwilkie) [inline probe/tag/origin_host_tagger_test.go]: Yes unfortunately some taggers might fail, so all have to be able go.  Pike!

**Comment by paulbellamy** [comment]
LGTM, then

**Pull request merged**